File: Src/java/elm/src/jvmTest/java/org/cqframework/cql/elm/TestUtils.java
package org.cqframework.cql.elm;

import org.cqframework.cql.cql2elm.CqlCompilerOptions;
import org.cqframework.cql.cql2elm.CqlTranslator;
import org.cqframework.cql.cql2elm.LibraryManager;
import org.cqframework.cql.cql2elm.ModelManager;
import org.hl7.cql.model.NamespaceInfo;

import java.io.IOException;
import java.io.InputStream;

import static kotlinx.io.CoreKt.buffered;
import static kotlinx.io.JvmCoreKt.asSource;

public class TestUtils {
    public static CqlTranslator createTranslatorFromStream(String testFileName, CqlCompilerOptions.Options... options)
            throws IOException {
        return createTranslatorFromStream(null, testFileName, options);
    }

    public static CqlTranslator createTranslatorFromStream(
            NamespaceInfo namespaceInfo, String testFileName, CqlCompilerOptions.Options... options)
            throws IOException {
        InputStream inputStream = TestUtils.class.getResourceAsStream(testFileName);
        return createTranslatorFromStream(null, inputStream, options);
    }

    public static CqlTranslator createTranslatorFromStream(
            InputStream inputStream, CqlCompilerOptions.Options... options) throws IOException {
        return createTranslatorFromStream(null, inputStream, options);
    }

    public static CqlTranslator createTranslatorFromStream(
            NamespaceInfo namespaceInfo, InputStream inputStream, CqlCompilerOptions.Options... options)
            throws IOException {
        ModelManager modelManager = new ModelManager();
        var compilerOptions = new CqlCompilerOptions(options);
        LibraryManager libraryManager = new LibraryManager(modelManager, compilerOptions);
        libraryManager.getLibrarySourceLoader().registerProvider(new TestLibrarySourceProvider());
        CqlTranslator translator = CqlTranslator.fromSource(namespaceInfo, buffered(asSource(inputStream)), libraryManager);
        return translator;
    }
}

File: Src/java/elm/src/jvmTest/kotlin/org/cqframework/cql/elm/TestUtils.kt
package org.cqframework.cql.elm

import java.io.IOException
import java.io.InputStream
import kotlinx.io.asSource
import kotlinx.io.buffered
import org.cqframework.cql.cql2elm.CqlCompilerOptions
import org.cqframework.cql.cql2elm.CqlTranslator
import org.cqframework.cql.cql2elm.LibraryManager
import org.cqframework.cql.cql2elm.ModelManager
import org.hl7.cql.model.NamespaceInfo

object TestUtils {
    @Throws(IOException::class)
    fun createTranslatorFromStream(
        testFileName: String,
        vararg options: CqlCompilerOptions.Options
    ): CqlTranslator {
        return createTranslatorFromStream(null, testFileName, *options)
    }

    @Throws(IOException::class)
    fun createTranslatorFromStream(
        namespaceInfo: NamespaceInfo?,
        testFileName: String,
        vararg options: CqlCompilerOptions.Options
    ): CqlTranslator {
        val inputStream = TestUtils::class.java.getResourceAsStream(testFileName)
        return createTranslatorFromStream(namespaceInfo, inputStream!!, *options)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun createTranslatorFromStream(
        inputStream: InputStream,
        vararg options: CqlCompilerOptions.Options
    ): CqlTranslator {
        return createTranslatorFromStream(null, inputStream, *options)
    }

    @Throws(IOException::class)
    fun createTranslatorFromStream(
        namespaceInfo: NamespaceInfo?,
        inputStream: InputStream,
        vararg options: CqlCompilerOptions.Options
    ): CqlTranslator {
        val modelManager = ModelManager()
        val compilerOptions = CqlCompilerOptions(*options)
        val libraryManager = LibraryManager(modelManager, compilerOptions)
        libraryManager.librarySourceLoader.registerProvider(TestLibrarySourceProvider())
        val translator =
            CqlTranslator.fromSource(
                namespaceInfo,
                inputStream.asSource().buffered(),
                libraryManager
            )
        return translator
    }
}
